[HW2] (SOLID & GRASP)

AlwaysSayNo opened the pull request:
Before PR was created, most of the structures in our packages already had SOLID principles implemented in them: 
- SRP - services, jobs, controllers, etc;
- DIP - services, jobs, controllers, etc.

Environment variable logic for configuration and schedulers were redesigned to support DIP.
As for the GRASP principles, the project was already following them before PR.

Additionally, work was done to improve the docker-compose logic. Previously, the program code was responsible for retries when connecting to the database (waitDBIsUp) and migration to the database (RunMigrations). All this logic was moved to the docker-compose file.

Tnx @andriihar and @AndrewSergienko for your useful comments!

Commit:
[SOLID] (Updated config logic)

- Updated config logic to effectively use env files and environment variables. Now viper is used as source of truth both with file/os variables.
- Separated configs in different structures to have Dependency Inversion Principle for this logic and runtime checks for missing configs.

Commit:
[SOLID] (Updated schedulers; Refactoring)

- Now we create a cron scheduler in job.go StartAllJobs function and pass it in other jobs, instead of creating this object in the jobs explicitly. This way we avoid violation of DIP.
- Renamed controller register functions, so they have verbs in the beginning.

Commit:
[Refactoring] (Updated docker-compose)

- Updated docker-compose using healthcheck, so application could properly wait until database is started.
- Updated docker-compose using commands, so docker could make migrations itself.
- All these needed to avoid code manipulation on Go side and use existing tools.
- Updated README.md file to support new logic.

Commit:
[SOLID] (Interfaces for services and controllers)

- Added interfaces and services  for all controllers and services.

andiserg commented on internal/service/currency-service.go:
```diff
@@ -16,44 +16,50 @@ import (
 )
 
-type CurrencyService struct {
+type CurrencyService interface {
+	GetCurrencyInfo() dto.CurrencyInfoDTO
+	GetCurrencyRate() dto.CurrencyResponseDto
+	UpdateCurrencyRates() error
+}
```
interfaces must be at the application layer, that is, they belong to a client and not to an implementation

andiserg replied:
This comment also applies to other services

AlwaysSayNo replied:
Moved interfaces closer to their clients. Since some services are used among different packages, a decision was made to create a separate folder for interfaces in ./pkg. 
![image](https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-AlwaysSayNo/assets/98183720/c48e6430-0e63-4a21-89af-9cffbe648eeb)

andriihar replied:
In Go, you don`t need a separate package for interfaces.

andiserg commented on pkg/controller/currency-controller.go:
```diff
@@ -7,18 +7,22 @@ import (
 
-type CurrencyController struct {
-	currencyService *service.CurrencyService
+type CurrencyController interface {
+	GetLatest(ctx *gin.Context)
```
Who is a client of this interface? This intefrace breaks the DIP because it depends on infractructure details like the gin framework.

andiserg replied:
You don't need to define controller interfaces, because the direction of dependencies is not broken here: the web framework calls controllers; controllers call services and return the result.

Controllers do not depend on services because there are interfaces. But I don't know about the dependence of a web framework on controllers. A web controller is a framework-specified object that must receive certain arguments and return a framework-defined result. You can define independent controllers that receive and return their own objects, but I don't see the point in this, because a controller should be a simple entry point object without complex logic, so this solution only complicates the project structure.

andiserg replied:
This comment also applies to other controllers

AlwaysSayNo replied:
Totaly agree with your comments. I made these interfaces since for testing golang requires them. But quickly realized that they are redundant for testing.
Removed these interfaces.

Commit:
[Interfaces] (Changes after comments)

- Moved service interfaces closer to clients into ./pkg/interface/ folder.
- Deleted redundant controller interfaces.

Commit:
[Interfaces] (Changes after comments)

- Moved service interfaces closer to clients into ./pkg/interface/ folder.
- Deleted redundant controller interfaces.

andriihar commented on internal/job/currency-job.go:
```diff
@@ -2,17 +2,16 @@ package job
-func UpdateCurrencyJob(currencyService *service.CurrencyService) {
-	scheduler := cron.New(cron.WithLocation(time.UTC))
-	_, err := scheduler.AddFunc("0 * * * *", func() {
+func UpdateCurrencyJob(cron *cron.Cron, currencyService service.CurrencyService) {
```
As for me service.CurrencyService is not very good name for interface. And also, as @AndrewSergienko was mentioned https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-AlwaysSayNo/pull/3#discussion_r1643902395

Interfaces is  better to locate near it is used.

Something like this

```go
type CurrencyUpdater interface {
	UpdateCurrencyRates() error
}

// UpdateCurrencyJob is a cron function to update currency service cache.
// It is executed every hour.
func UpdateCurrencyJob(cron *cron.Cron, currencyService CurrencyUpdater) {
	_, err := cron.AddFunc("0 * * * *", func() {
		log.Println("Start job: Update Currency Rates")

		if err := currencyService.UpdateCurrencyRates(); err != nil {
			log.Printf("error with: Update Currency Rates - %v\n", err)
		} else {
			log.Println("Finish job: Update Currency Rates")
		}
	})
	if err != nil {
		log.Printf("failed to start UpdateCurrencyJob scheduler: %v\n", err)
		return
	}
}
```

You just need only one method `UpdateCurrencyRates` not all methods from interface

AlwaysSayNo replied:
Done.

andriihar commented on internal/job/send-emails-job.go:
```diff
@@ -2,18 +2,16 @@ package job
-	scheduler := cron.New(cron.WithLocation(time.UTC))
-
-	_, err := scheduler.AddFunc("0 9 * * *", func() {
+func SendEmailsJob(cron *cron.Cron, emailService service.EmailService) {
```
Here the same **`Go interfaces generally belong in the package that uses values of the interface type, not the package that implements those values`**

https://go.dev/wiki/CodeReviewComments#interfaces

```go
type EmailSender interface {
	Send() error
}
```

AlwaysSayNo replied:
Done.

andriihar commented on internal/service/currency-service.go:
```diff
@@ -9,49 +9,52 @@ import (
 )
 
-type CurrencyService struct {
+type CurrencyServiceImpl struct {
```
And after, if you name interface `CurrencyUpdater` you can name this `CurrencyService`. 
If you put interface in the package that uses values of the interface type, you will not have name collision.

andriihar commented on internal/service/currency-service.go:
```diff
@@ -9,49 +9,52 @@ import (
 	// A cache value for 3rd party API response.
 	var currencyInfo dto.CurrencyInfoDTO
-	c := &CurrencyService{
+	c := &CurrencyServiceImpl{
```
As for me, this func violates S (single responsibility from SOLID) 

* create instance
* update rates

This is constructor, 


```go

type CurrencyServiceImpl struct {
	currencyInfo dto.CurrencyInfoDTO
	cnf          config.CurrencyServiceConfig
}

func NewCurrencyServiceImpl(currencyInfo dto.CurrencyInfoDTO, cnf config.CurrencyServiceConfig) *CurrencyServiceImpl {
	return &CurrencyServiceImpl{currencyInfo: currencyInfo, cnf: cnf}
}

```

AlwaysSayNo replied:
Fixed this issue in the future PR HW3.

andriihar commented on internal/service/email-service.go:
```diff
@@ -21,23 +22,28 @@ type CurrencyEmailData struct {
-type EmailService struct {
-	userService     *UserService
-	currencyService *CurrencyService
+type EmailServiceImpl struct {
```
Same as in Currency

AlwaysSayNo replied:
Done.

andriihar commented on pkg/interface/service/currency-service.go:
```diff
@@ -0,0 +1,9 @@
+
+import "genesis-currency-api/pkg/dto"
+
+type CurrencyService interface {
```
https://go.dev/wiki/CodeReviewComments#interfaces

Go interfaces generally belong in the package that uses values of the interface type, not the package that implements those values

AlwaysSayNo replied:
Done.

andriihar commented on internal/service/user-service.go:
```diff
@@ -37,7 +37,7 @@ func (s *UserService) Save(user dto.UserSaveRequestDTO) (dto.UserResponseDTO, er
 // GetAll is used to get all available in database users' information.
 // Returns all available UserResponseDTO.
-func (s *UserService) GetAll() ([]dto.UserResponseDTO, error) {
+func (s *UserServiceImpl) GetAll() ([]dto.UserResponseDTO, error) {
```
Is it a repository?

AlwaysSayNo replied:
I'll check repository pattern and probably change this.

andriihar commented on pkg/controller/currency-controller.go:
```diff
@@ -3,22 +3,23 @@ package controller
-func CurrencyRegisterRoutes(r *gin.Engine, s *service.CurrencyService) {
-	c := &CurrencyController{
+// RegisterCurrencyRoutes creates an instance of CurrencyControllerImpl and registers routes for it.
+func RegisterCurrencyRoutes(r *gin.Engine, s service.CurrencyService) {
```
Look like this func also violates `S`. 

Is create an instance and register routes. 

Also, I prefere to move this to another package (api) in internal directory. Create constructor to all handlers (controllers) 

In out projects we use struct like this
<img width="292" alt="image" src="https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-AlwaysSayNo/assets/25195930/031a5ba9-e7b3-4f01-a98d-b57182071b71">

AlwaysSayNo replied:
Done.

andriihar commented on pkg/controller/user-controller.go:
```diff
@@ -3,18 +3,19 @@ package controller
-type UserController struct {
-	userService  *service.UserService
-	emailService *service.EmailService
+type UserControllerImpl struct {
```
Here needs its own interface.

AlwaysSayNo replied:
Done.

Commit:
[Interfaces] (Interfaces in jobs)

- Created interfaces in jobs.

Commit:
[Interfaces] (Handler refactoring)

- Refactored handlers: moved them into separate package in internal package.
- Now we have functions to create handlers + a function to register routes.
- Created interfaces for handlers attributes.
